### DISCUSSION TITLE
Defining schema for coordinates

### DISCUSSION THREAD

**Discussion by marko-asplund** [SystemFw/dynosaur#14] [Q&A] [answered]
I'm trying to define schema for representing coordinates. It's an existing data model that looks like this:
```
"L": [
  { "N": "24.92887263472079" },
  { "N": "60.17018555542941" }
]
```

The following declaration seems to be working, but is a bit verbose.
Could this be simplified somehow?

```
def dvToDouble(d: Option[DynamoValue], msg: String) = ???

val pointSchema: Schema[Point] = Schema[DynamoValue].xmap { f =>
  val res = for {
    list <- Either.fromOption(f.l, "missing or empty coordinates list")
    x    <- dvToDouble(list.get(0), "invalid or missing x coordinate")
    y    <- dvToDouble(list.get(1), "invalid or missing y coordinate")
  } yield geometryFactory.createPoint(new Coordinate(x, y))
  res.leftMap(s => Schema.ReadError(s))
}(p => Either.right(dv.l(dv.n(p.getX), dv.n(p.getY))))
```

**Comment by SystemFw** [comment, marked as answer]
The AttributeValue is slightly weird there, so I expect some friction. Couple of thoughts:
- You can probably use `Schema[List[Double]]` as a starting point, which should save some code
- You can use `imapErr`since encoding doesn't fail.
- If the list only has one value, how do you know (logically) whether `x` or `y` are missing?

I haven't tested it, but something like:
```scala
val coordSchema: Schema[Coordinate] =
  Schema[List[Double]].imapErr {
    case List(x, y) => Right(Coordinate(x, y))
    case _ => Left(ReadError("wrong coordinates")
  } { coord => List(coord.x, coord.y) }
```

**Reply by marko-asplund** [reply]
Just tried this out and seems to work nicely, in addition to being less verbose - thanks! 🙇
